src/ParseNode.js

```diff
@@ -31,3 +34,239 @@ export default class ParseNode {
         this.loc = SourceLocation.range(first, last);
     }
 }
+
+export type NodeType = $Keys<ParseNodeTypes>;
+export type NodeValue<TYPE: NodeType> = $ElementType<ParseNodeTypes, TYPE>;
+
+export type AccentStructType = {
+    type: "accent",
+    label: string | {
+        text: string,
+        math?: string,
+    },
+    isStretchy: boolean,
+    isShifty: boolean,
+    base: ParseNode<*>,
+};
+
+// Map from `type` field value to corresponding `value` type.
+type ParseNodeTypes = {
+    "array": ArrayEnvNodeData,
+    "accent": string | AccentStructType,
+    "color": {|
+        type: "color",
+        color: string,
+        value: ParseNode<*>[],
+    |},
+    "leftright": {|
+        body: [{|
+            type: "array",
+            hskipBeforeAndAfter: boolean,
+        |} | ParseNode<*>],
+        left: string,
+        right: string,
+    |},
+    "op": string | {|
+        type: "op",
+        limits: boolean,
+        symbol: boolean,
+        alwaysHandleSupSub?: boolean,
+        body?: string,
+        value?: ParseNode<*>[],
+    |},
+    "ordgroup": ParseNode<*>[],
+    "size": {|
+        number: number,
+        unit: string,
+    |},
+    "styling": {|
+        type: "styling",
+        style: StyleStr,
+        value: ParseNode<*>[],
+    |},
+    "supsub": {|
+        base: ParseNode<*>,
+        sup?: ?ParseNode<*>,
+        sub?: ?ParseNode<*>,
+    |},
+    "text": {|
+        type: "text",
+        body: ParseNode<*>[],
+        font?: string,
+    |},
+    "textord": string,
+    "url": string,
+    "verb": {|
+        body: string,
+        star: boolean,
+    |},
+    // From symbol groups, constructed in Parser.js via `symbols` lookup.
+    // ("accent" and "op" handled above to capture additional variants.)
+    "bin": string,
+    "close": string,
+    "inner": string,
+    "mathord": string,
+    "open": string,
+    "punct": string,
+    "rel": string,
+    "spacing": string,
+    "textord": string,
```

Comment:
All of the groups typed as a `string` stick out a like a sore thumb.  No action required on this, but something we'll probably want to refactor in the future.